File: s13_WorkManager/WorkManager/app/src/main/java/com/omeryildizce/workmanager/MainActivity.java
package com.omeryildizce.workmanager;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.work.Constraints;
import androidx.work.Data;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;
import androidx.work.WorkRequest;

import android.annotation.SuppressLint;
import android.os.Bundle;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {

    @SuppressLint("EnqueueWork")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Data data = new Data.Builder().putInt("intKey", 1).build();
        Constraints constraints = new Constraints.Builder()
                //.setRequiredNetworkType(NetworkType.CONNECTED)
                .setRequiresBatteryNotLow(true)
                .setRequiresCharging(false)
                .build();
        /*
        WorkRequest workRequest = new OneTimeWorkRequest.Builder(RefreshDaatabase.class)
                .setConstraints(constraints)
                .setInputData(data)
                .setInitialDelay(5, TimeUnit.SECONDS)
                .addTag("myTag")
                .build();

        WorkManager.getInstance(this).enqueue(workRequest);
        */

        WorkRequest workRequest = new PeriodicWorkRequest.Builder(RefreshDaatabase.class, 1, TimeUnit.MINUTES )
                .setConstraints(constraints)
                .setInputData(data)
                .build();

        WorkManager.getInstance(this).enqueue(workRequest);
        WorkManager.getInstance(this).getWorkInfoByIdLiveData(workRequest.getId()).observe(this, new Observer<WorkInfo>() {
            @Override
            public void onChanged(WorkInfo workInfo) {
                if (workInfo.getState() == WorkInfo.State.RUNNING){
                    System.out.println("running");
                } else if (workInfo.getState() == WorkInfo.State.BLOCKED) {
                    System.out.println("blocked");
                    
                } else if (workInfo.getState() == WorkInfo.State.SUCCEEDED) {
                    System.out.println("succeeded");
                } else if (workInfo.getState() == WorkInfo.State.FAILED) {
                    System.out.println("failed");
                }
            }
        });
        // WorkManager.getInstance(this).cancelAllWork();

        // Chaining
        /*
        OneTimeWorkRequest oneTimeWorkRequest = new OneTimeWorkRequest.Builder(RefreshDaatabase.class)
                .setInputData(data)
                .setConstraints(constraints)
                .build();

        WorkManager.getInstance(this).beginWith(oneTimeWorkRequest)
                .then(oneTimeWorkRequest)
                .then(oneTimeWorkRequest)
                .enqueue();
        */
    }

}

File: s26_WorkManager/WorkManager/app/src/main/java/com/omeryildizce/workmanager/MainActivity.kt
package com.omeryildizce.workmanager

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.work.*
import com.omeryildizce.workmanager.databinding.ActivityMainBinding
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        val data = Data.Builder().putInt("intKey", 1).build()

        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .setRequiresCharging( false)
            .setRequiresBatteryNotLow(true)
            .build()

        /*
        val myWorkRequest: WorkRequest = OneTimeWorkRequestBuilder<RefreshDatabase>()
            .setConstraints(constraints)
            .setInputData(data)
            .setInitialDelay(30, TimeUnit.SECONDS)
            .addTag("myTag")
            .build()

        WorkManager.getInstance(this).enqueue(myWorkRequest)

         */

        val myRequest: PeriodicWorkRequest = PeriodicWorkRequestBuilder<RefreshDatabase>(1, TimeUnit.MINUTES)
            .setConstraints(constraints)
            .setInputData(data)
            .build()

        WorkManager.getInstance(this).enqueue(myRequest)
        WorkManager.getInstance(this).getWorkInfoByIdLiveData(myRequest.id)
            .observe(this, Observer {
                if (it.state == WorkInfo.State.RUNNING){
                    println("running")
                }else if(it.state == WorkInfo.State.FAILED)
                {
                    println("failed")
                }else if (it.state == WorkInfo.State.SUCCEEDED){
                    println("succeeded")
                }
            })


        // Chaining
        val oneTimeRequest = OneTimeWorkRequestBuilder<RefreshDatabase>()
            .setConstraints(constraints)
            .setInputData(data)
            .build()

        WorkManager.getInstance(this).beginWith(oneTimeRequest)
            .then(oneTimeRequest)
            .then(oneTimeRequest)
            .enqueue()

    // WorkManager.getInstance(this).cancelAllWork()
    }
}